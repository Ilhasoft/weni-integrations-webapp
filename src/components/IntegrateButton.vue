<template>
  <div>
    <LoadingButton
      ref="button"
      type="secondary"
      loadingPosition="center"
      :size="size"
      :iconCenter="icon"
      :isLoading="loadingCreateApp"
      :disabled="disabled"
      :text="text"
      @clicked="addApp(app)"
    />

    <add-modal ref="addModal" />
    <config-pop-up ref="configPopUp" />
  </div>
</template>

<script>
  import addModal from './AddModal.vue';
  import configPopUp from './config/ConfigPopUp.vue';
  import { unnnicCallAlert } from '@weni/unnnic-system';
  import { mapActions, mapState } from 'vuex';
  import LoadingButton from './LoadingButton.vue';
  import getEnv from '../utils/env';

  export default {
    name: 'IntegrateButton',
    components: { configPopUp, addModal, LoadingButton },
    props: {
      app: {
        type: Object,
        default: /* istanbul ignore next */ () => {},
      },
      icon: {
        type: String,
        default: null,
      },
      disabled: {
        type: Boolean,
        default: false,
      },
      text: {
        type: String,
        default: null,
      },
      size: {
        type: String,
        default: 'small',
      },
    },
    data() {
      return {
        loadingCreateApp: false,
        hasFBLoginList: ['wpp-cloud', 'ig', 'fba'],
        appCodeToFBAppId: {
          'wpp-cloud': getEnv('VUE_APP_WHATSAPP_FACEBOOK_APP_ID'),
          ig: getEnv('VUE_APP_FACEBOOK_APP_ID'),
        },
      };
    },
    created() {
      window.openWACloudPopUp = this.openWACloudPopUp;
    },
    computed: {
      ...mapState({
        createAppResponse: (state) => state.appType.createAppResponse,
        errorCreateApp: (state) => state.appType.errorCreateApp,
        project: (state) => state.auth.project,
      }),
    },
    methods: {
      ...mapActions(['createApp']),
      async addApp(app) {
        if (this.hasFBLoginList.includes(app.code) || app.config_design === 'pre-popup') {
          this.$refs.configPopUp.openPopUp(app);
          return;
        }

        let code = app.code;
        const payload = {
          project_uuid: this.project,
        };
        if (app.generic) {
          code = 'generic';
          payload.channel_code = app.code;
        }
        this.loadingCreateApp = true;
        await this.createApp({ code, payload });
        this.loadingCreateApp = false;

        if (this.errorCreateApp) {
          this.callErrorModal({ text: this.$t('apps.details.status_error') });
          return;
        }

        if (app.config_design === 'popup') {
          app.config = this.createAppResponse.config;
          this.$refs.configPopUp.openPopUp(app);
        } else {
          this.$refs.addModal.toggleModal();
        }
        this.$emit('update');
      },
      openWACloudPopUp(app, input_token) {
        const customData = {
          input_token,
        };
        this.$refs.configPopUp.openPopUp(app, customData);
      },
      callErrorModal({ text }) {
        unnnicCallAlert({
          props: {
            text,
            title: this.$t('general.error'),
            icon: 'check-circle-1-1',
            scheme: 'feedback-red',
            position: 'bottom-right',
            closeText: this.$t('general.Close'),
          },
          seconds: 6,
        });
      },
    },
  };
</script>
<style></style>
