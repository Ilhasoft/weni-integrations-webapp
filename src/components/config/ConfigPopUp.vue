<template>
  <div class="config-popup" v-if="show">
    <component
      class="config-popup__component"
      :is="currentComponent"
      :v-bind="$attrs"
      :app="currentApp"
      :customData="currentCustomData"
      @closePopUp="closePopUp"
    />
  </div>
</template>

<script>
  import wppDemoConfig from './channels/wpp_demo/Config.vue';
  import wppPageSelection from './channels/whatsapp/PageSelection.vue';

  export default {
    name: 'Config-PopUp',
    data() {
      return {
        show: false,
        type: '',
        currentApp: {},
        currentCustomData: null,
      };
    },
    methods: {
      closePopUp() {
        this.show = false;
      },
      openPopUp(app, customData) {
        this.type = app.code;
        this.currentApp = app;
        this.currentCustomData = customData;
        this.show = true;
      },
    },
    computed: {
      currentComponent() {
        if (this.type === 'wpp-demo') return wppDemoConfig;
        if (this.type === 'wpp') return wppPageSelection;
        return wppDemoConfig;
      },
    },
  };
</script>

<style lang="scss" scoped></style>
