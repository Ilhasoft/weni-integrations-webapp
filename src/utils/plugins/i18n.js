import Vue from 'vue';
import VueI18n from 'vue-i18n';

Vue.use(VueI18n);

import en from '../../locales/en.json';
import pt_br from '../../locales/pt_br.json';
import es_es from '../../locales/es_es.json';

const languages = {
  'en-us': en,
  'pt-br': pt_br,
  es: es_es,
};

const messages = Object.assign(languages);

const i18n = new VueI18n({
  locale: 'en-us',
  fallbackLocale: 'en-us',
  messages,
});

export default i18n;
