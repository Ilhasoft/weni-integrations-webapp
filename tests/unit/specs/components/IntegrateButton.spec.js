jest.mock('@/api/appType', () => {
  return {
    createApp: jest.fn(),
  };
});

import { unnnicCallAlert as mockUnnnicCallAlert } from '@weni/unnnic-system';

jest.mock('@weni/unnnic-system', () => ({
  ...jest.requireActual('@weni/unnnic-system'),
  unnnicCallAlert: jest.fn(),
}));

import VueRouter from 'vue-router';
import Vuex from 'vuex';
import { mount, createLocalVue } from '@vue/test-utils';
import IntegrateButton from '@/components/IntegrateButton.vue';
import addModal from '@/components/AddModal.vue';
import ConfigPopUp from '@/components/config/ConfigPopUp.vue';
import i18n from '@/utils/plugins/i18n';
import { singleApp } from '../../../__mocks__/appMock';

const router = new VueRouter();

const localVue = createLocalVue();
localVue.use(VueRouter);
localVue.use(Vuex);

describe('IntegrateButton.vue', () => {
  let wrapper, actions, getters, state, store;

  beforeEach(() => {
    actions = {
      createApp: jest.fn(() => {}),
    };

    getters = {
      getSelectedProject: jest.fn(() => {
        return '123';
      }),
    };

    state = {
      appType: {
        createAppResponse: null,
        errorCreateApp: false,
      },
    };

    store = new Vuex.Store({
      actions,
      getters,
      state,
    });

    wrapper = mount(IntegrateButton, {
      localVue,
      store,
      i18n,
      router,
      mocks: {
        $t: () => 'some specific text',
      },
      stubs: {
        UnnnicButton: true,
        UnnnicModal: true,
        ConfigPopUp,
        addModal,
      },
      propsData: {
        app: singleApp,
        icon: 'add-1',
        disabled: false,
        text: 'Add',
        size: 'large',
      },
    });
  });

  it('should be rendered properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call addApp on button click', async () => {
    const spy = spyOn(wrapper.vm, 'addApp');
    expect(spy).not.toHaveBeenCalled();

    const button = wrapper.findComponent({ ref: 'button' });
    button.vm.$emit('clicked', { stopPropagation: () => {} });

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith(wrapper.props('app'));
  });

  describe('created()', () => {
    it('should add fetchDebugToken into DOM', async () => {
      expect(window.fetchDebugToken).toEqual(wrapper.vm.fetchDebugToken);
    });

    it('should add fetchPhoneNumbers into DOM', async () => {
      expect(window.fetchPhoneNumbers).toEqual(wrapper.vm.fetchPhoneNumbers);
    });
  });

  describe('addApp()', () => {
    it('should call facebookLogin if app code is wpp-cloud', async () => {
      const spy = spyOn(wrapper.vm, 'facebookLoginAppCreation');
      expect(spy).not.toHaveBeenCalled();
      const app = {
        code: 'wpp-cloud',
      };
      await wrapper.vm.addApp(app);
      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('should call createApp method', async () => {
      expect(actions.createApp).not.toHaveBeenCalled();
      const app = {
        code: 'code',
        config_design: 'sidemenu',
      };
      await wrapper.vm.addApp(app);
      expect(actions.createApp).toHaveBeenCalledTimes(1);
      expect(actions.createApp).toHaveBeenCalledWith(expect.any(Object), {
        code: app.code,
        payload: expect.any(Object),
      });
    });

    it('should call unnnicCallAlert on error', async () => {
      store.state.appType.errorCreateApp = true;
      const spy = spyOn(wrapper.vm, 'callErrorModal');
      expect(spy).not.toHaveBeenCalled();
      const app = {
        code: 'code',
        config_design: 'sidemenu',
      };
      await wrapper.vm.addApp(app);
      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('should call toggleModal if design not popup', async () => {
      const spy = spyOn(wrapper.vm.$refs.addModal, 'toggleModal');
      expect(spy).not.toHaveBeenCalled();
      const app = {
        code: 'code',
        config_design: 'sidemenu',
      };
      await wrapper.vm.addApp(app);
      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('should call openPopUp if design is popup', async () => {
      const spy = spyOn(wrapper.vm.$refs.configPopUp, 'openPopUp');
      const appConfig = {
        config: {
          redirect_url: 'https://url.com',
        },
      };
      actions.createApp.mockImplementation(() => {
        return Promise.resolve({
          data: appConfig,
        });
      });
      store.state.appType.createAppResponse = appConfig;
      expect(spy).not.toHaveBeenCalled();
      const app = {
        code: 'code',
        config_design: 'popup',
      };
      await wrapper.vm.addApp(app);
      expect(spy).toHaveBeenCalledTimes(1);
    });
  });

  describe('openWACloudPopUp()', () => {
    it('should call openPopUp with parameter token', () => {
      const spy = spyOn(wrapper.vm.$refs.configPopUp, 'openPopUp');
      expect(spy).not.toHaveBeenCalled();
      const token = '123';
      wrapper.vm.openWACloudPopUp(wrapper.vm.app, token);
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(wrapper.vm.app, { input_token: token });
    });
  });

  describe('callErrorModal', () => {
    it('should call unnnicCallAlert', () => {
      expect(mockUnnnicCallAlert).not.toHaveBeenCalled();
      wrapper.vm.callErrorModal({ text: 'error text' });
      expect(mockUnnnicCallAlert).toHaveBeenCalledTimes(1);
      expect(mockUnnnicCallAlert).toHaveBeenCalledWith(
        expect.objectContaining({
          props: {
            text: 'error text',
            title: expect.any(String),
            icon: expect.any(String),
            scheme: expect.any(String),
            position: expect.any(String),
            closeText: expect.any(String),
          },
          seconds: expect.any(Number),
        }),
      );
    });
  });
});
