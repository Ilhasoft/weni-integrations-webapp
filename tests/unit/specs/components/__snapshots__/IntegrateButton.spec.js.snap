// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IntegrateButton.vue should be rendered properly 1`] = `
<div>
  <unnnicbutton-stub size="large" type="secondary" iconcenter="add-1">
    Add
  </unnnicbutton-stub>
  <unnnicmodal-stub text="some specific text" modalicon="check-circle-1-1" closeicon="true" scheme="feedback-green" class="add-modal"><span>some specific text</span>
    <unnnicbutton-stub size="large" type="terciary">some specific text</unnnicbutton-stub>
    <unnnicbutton-stub size="large" type="primary">
      some specific text
    </unnnicbutton-stub>
  </unnnicmodal-stub>
  <!---->
</div>
`;
